Sources/BSON/BSONObjectID.swift

```diff
@@ -0,0 +1,89 @@
+import Foundation
+import NIO
+import NIOConcurrencyHelpers
+
+/// A struct to represent the BSON ObjectID type.
+public struct BSONObjectID: Equatable, Hashable, CustomStringConvertible {
+    /// This `BSONObjectID`'s data represented as a `String`.
+    public var hex: String { self.oid.reduce("") { $0 + String(format: "%02x", $1) } }
+
+    public var description: String { self.hex }
+
+    private let oid: [UInt8]
+    private static let generator = ObjectIDGenerator()
+
+    /// Initializes a new `BSONObjectID`.
+    public init() {
+        self.oid = Self.generator.generate()
+    }
+
+    /// Initializes a new `BSONObjectID`.
+    internal init(_ bytes: [UInt8]) {
+        self.oid = bytes
+    }
+
+    /// Initializes an `BSONObjectID` from the provided hex `String`.
+    /// - Throws:
+    ///   - `BSONError.InvalidArgumentError` if string passed is not a valid BSONObjectID
+    /// - SeeAlso: https://github.com/mongodb/specifications/blob/master/source/objectid.rst
+    public init(_ hex: String) throws {
+        guard hex.utf8.count == 24 else {
```

Comment:
instead of a magic number, we should define a static constant under `ObjectID` that is it's length (i.e. 12). Then we can do 2 * that constant here, and the local variable `len` can just go away.